### ISSUE TITLE
Wesnoth can't be compiled with Xcode 10 Beta

### ISSUE THREAD

**Issue by hrubymar10** [wesnoth/wesnoth#3350]
# Description
Wesnoth can't be compiled with Apple's newest betas of Xcode. It is probably issue just with Xcode not whole OS. I believe that it is also reproducible with stable version of macOS with newest Xcode beta.

This problem affects all current wesnoth branches.

This problem is reproducible since first Xcode 10 Beta.

# OS Info
**OS:** macOS Mojave 10.14 (18A336e) Beta 4
**Xcode:** Xcode 10.0 (10L213o) Beta 4
**Clang:**
```
Apple LLVM version 10.0.0 (clang-1000.10.40.1)
Target: x86_64-apple-darwin18.0.0
Thread model: posix
InstalledDir: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

# Problem
During compiling xcode/clang generates around 350 compile errors.

Example errors (generated from `wesnoth/wesnoth:1.14` a21c4d7ce3d2acf122eb4170266f961740baa526:
```
/Users/user/xcode/wesnoth_1.14/src/serialization/preprocessor.hpp:93:15: Field has incomplete type 'version_info'
/Users/user/xcode/wesnoth_1.14/src/server/server.cpp:29:10: In file included from /Users/user/xcode/wesnoth_1.14/src/server/server.cpp:29:
/Users/user/xcode/wesnoth_1.14/src/game_config.hpp:18:7: Forward declaration of 'version_info'
```

```
/Users/user/xcode/wesnoth_1.14/src/serialization/preprocessor.cpp:1324:14: Use of undeclared identifier 'OP_INVALID'
```

```
/Users/user/xcode/wesnoth_1.14/src/preferences/game.cpp:673:9: Use of undeclared identifier 'settings'; did you mean 'gui2::settings'?
```

# Full log
[wesnoth_log.txt](https://github.com/wesnoth/wesnoth/files/2204929/wesnoth_log.txt)

**Comment by jyrkive** [comment]
Judging from the first example error, it looks like a bug in Apple Clang. `serialization/preprocessor.hpp` #includes `version.hpp` that contains the definition of `version_info`, but Clang is choosing the forward declaration that it saw first instead.

**Comment by CelticMinstrel** [comment]
Unless there's some sort of namespacing issue whereby the compiler thinks the forward declaration and the definition somehow ended up in different namespaces?

**Comment by hrubymar10** [comment]
It is probably issue with **Always Search User Paths** option which is depracated since Xcode 8 or 9

![2018-08-08 at 15 50](https://user-images.githubusercontent.com/2218817/43841178-cfa8fd0a-9b22-11e8-9ffc-bbed4745bb1a.png)

Let's take for example this line: https://github.com/wesnoth/wesnoth/blob/355c92ba8331e74f1b64e6eb3266e581b0b6fc97/src/addon/info.hpp#L17

I can fix this issue by including full path: `#include "../version.hpp"`
or by this: `#include <version.hpp>`

**Comment by GregoryLundberg** [comment]
Don't use <> version.hpp is not a library-thing, it's a project-thing.

I tried to clean up a bunch of crap like this a year or two ago but it creeps back in over time.

**Comment by CelticMinstrel** [comment]
From the sounds of it, we may need to explicitly add the "src" directory as an include path.

**Comment by GregoryLundberg** [comment]
We already do.

The problem is that what it says is "version.hpp" and what it should say "addon/version.hpp" or "../version.hpp".

Personally, I prefer "addon/version.hpp" so we can move the source file elsewhere and it can still find the correct header, and, if we move the header, we can more accurately grep for it (since "../version.hpp" might mean "addon/version.hpp" or it might mean "my/junky/patch/version.hpp")

**Comment by hrubymar10** [comment]
From screenshot you can see that `src` folder is already included:
![2018-08-09 at 15 52](https://user-images.githubusercontent.com/2218817/43903270-5ac2dd2e-9bec-11e8-91f3-3db7cecebee5.png)

Problem is that there are 2 `version.hpp` files. One in root of src (`wesnoth_working_dir/src/version.hpp`) and the other in addon dir (`wesnoth_working_dir/src/addon/version.hpp`).

This include:
https://github.com/wesnoth/wesnoth/blob/355c92ba8331e74f1b64e6eb3266e581b0b6fc97/src/addon/info.hpp#L17

Should include `version.hpp` from root of src (`wesnoth_working_dir/src/version.hpp`) but it includes version.hpp from addon dir (`wesnoth_working_dir/src/addon/version.hpp`). That's what problem is. Xcode 10 have this new bug/feature/whatever that it doesn't try user paths first. It first tried search inside addon dir.

**Comment by GregoryLundberg** [comment]
Yeah, I was cooking breakfast and thinking exactly that.

Best solution is to break things and change the less-used one to another name.

But ../ works, use it.

**Comment by hrubymar10** [comment]
I see 2 options from here now:

## 1.
Wait if Apple decides that this behaviour is not feature and will fix it as bug. If Apple won't fix it until oficial macOS release in around October, then I'll create PR where I'll replace all these includes to `../` ones.

## 2.
I'll create this PR asap and we won't care about Apple's decision.


I would like to see opinion of at least these people:
@GregoryLundberg , @jyrkive , @Vultraz , @CelticMinstrel

**Comment by jyrkive** [comment]
I'd prefer switching to the full relative paths immediately.

**Comment by GregoryLundberg** [comment]
The problem is that we have

"src/version.hpp"

and we have

"src/addons/version.hpp"

and we have a file in "/src/addons" which needs to include  both.

There is no standard C/C++ way to say "I want version.hpp, but not the one from the directory I came from.

So we can Unix-hack it using '../' or we can bite the bullet and fix it by stating there is a Wesnoth coding standard which states that any header name, such as version.hpp, which appears  directly in the src directory is a reserved header name and cannot be used elsewhere. (this IS the rule we have, btw) And, pursuant to rule, rename addons/version.hpp to something else such as addons/addon_version.hpp

**Comment by hrubymar10** [comment]
@GregoryLundberg this way will probably also work (I'll try it to be sure).

**Comment by GregoryLundberg** [comment]
The problem is, Apple's old defaults where brain-dead. The new defaults are much more logical, and break fewer things. But the change from "search the path ('src') before here ('.')" to the normal "search here before the path" did not break us, it exposed our bad choices. Rename the header in addons and we don't care what the compiler does, we're always telling it the right thing.

**Comment by hrubymar10** [comment]
We must include full path. I just tested that there is for example this include:
https://github.com/wesnoth/wesnoth/blob/355c92ba8331e74f1b64e6eb3266e581b0b6fc97/src/serialization/preprocessor.hpp#L27

And there is no `working_dir/src/serialization/version.hpp`. Only fix (so far) for include above is including full relative path `#include "../version.hpp"`.

**Comment by GregoryLundberg** [comment]
That would be a bug which should be reported to Apple. You told it where to look and it didn't. That is a standards-compliance regression.

**Comment by hrubymar10** [comment]
I open bug report ticket on https://bugreport.apple.com/
-----------
Bug Report ID: 43093884
```
Area:
Xcode

Summary:
I am developer of open-source game called The Battle for Wesnoth. Link to source code is here: https://github.com/wesnoth/wesnoth . With Xcode 9 our Xcode project (https://github.com/wesnoth/wesnoth/tree/master/projectfiles/Xcode) could be compiled without any problems. We have set 'Always Search User Paths' option to true. But since very first beta of Xcode 10 project couldn't be compiled (compile log: https://github.com/wesnoth/wesnoth/files/2204929/wesnoth_log.txt). I tried simulate Xcode 10's behaviour with Xcode 9 and I found that if set 'Always Search User Paths' option to false in Xcode 9 I get totally same problem. So that makes me think that Xcode 10 just ignores 'Always Search User Paths option' option at all.

Issue is also tracked here: https://github.com/wesnoth/wesnoth/issues/3350

Steps to Reproduce:
Use Xcode 10 Beta to compile our source code: https://github.com/wesnoth/wesnoth/tree/master/projectfiles/Xcode

You can use these maintained build steps: https://github.com/wesnoth/wesnoth/blob/master/projectfiles/Xcode/README.md

And I really recommend you follow 'package builds' steps.

Expected Results:
Build is sucessfull as in Xcode 9

Actual Results:
Build fails with Xcode 10 at all. If you use Xcode 9 and set 'Always Search User Paths' option to false you will get same result.

Version/Build:
It is reproducible since first developer beta of macOS Mojave and Xcode 10.

Problem was first reported with:
OS: macOS Mojave 10.14 (18A336e) Beta 4
Xcode: Xcode 10.0 (10L213o) Beta 4

In time of writing:
OS: macOS Mojave 10.14 (18A353d) Beta 6
Xcode: Xcode 10.0 (10L221o) Beta 5

Configuration:
Source Code: https://github.com/wesnoth/wesnoth
Project File: https://github.com/wesnoth/wesnoth/tree/master/projectfiles/Xcode
sysdiagnose attached
```

**Comment by CelticMinstrel** [comment]
I basically agree with Gregory, except I'd prefer to rename the toplevel version.hpp rather than the addons one.

In the case you cited in serialization, that include should work as-is as long as  "src" is listed as an include path, so I don't know what's going on there.

**Comment by hrubymar10** [comment]
# Response from Apple

**Apple Developer Relations**
August 10 2018, 4:50 PM
There are no plans to address this based on the following:  Always Search User Paths is not supported in Xcode’s new build system - the functionality of setting it to YES (using traditional headermaps) has been long deprecated and is not being implemented in the new build system. You will have to convert your project to work with the new behavior. There’s a warning about this in the build log.  The most common issue is that there may be headers you need to add to your project so they get added to the headermap, as the new mechanism does not scan your project directories for headers, it only considers headers in the project. If your project has multiple header files with the same name, then you should make sure each header is associated with the target it belongs to (at least by making it a ‘Project’ header in that target’s Copy Headers build phase) so it takes precedence for that target over other headers of the same name.  We are now closing this report.  If you have questions about the resolution, or if this is still a critical issue for you, then please update your bug report with that information.  Please be sure to regularly check new Apple releases for any updates that might affect this issue.

**Comment by jyrkive** [comment]
@hrubymar10 Could you check if `version.hpp` (one of the affected headers) is present in the Xcode project?

> If your project has multiple header files with the same name, then you should make sure each header is associated with the target it belongs to

That part sounds strange. It's as if only one header of each name is allowed for each target. Our `wesnoth` target frequently uses multiple headers with the same name...

**Comment by hrubymar10** [comment]
I just verified that `version.hpp` and `version.cpp` from `src` are included in Xcode project and are targeted from Wesnoth target

**Comment by jyrkive** [comment]
In that case I bet Xcode is choosing [desktop/version.hpp](https://github.com/wesnoth/wesnoth/blob/c10fc7f08a82c6965c2494a3341d39a0dff79d66/src/desktop/version.hpp) or [spirit_po/version.hpp](https://github.com/wesnoth/wesnoth/blob/c10fc7f08a82c6965c2494a3341d39a0dff79d66/src/spirit_po/version.hpp) instead.

Looks like we have to ensure that every single header has a unique name. :(

Damn you, Apple...

**Comment by hrubymar10** [comment]
I am preparing new branch with fix now. I have fixed it locally already.

List of problematic files:
`src/image.hpp` -> `src/image_general.hpp`
`src/version.hpp` -> `src/version_general.hpp`
`src/settings.hpp` -> `src/settings_general.hpp`
`src/tests/utils/game_config_manager.hpp` -> `src/tests/utils/game_config_manager_tests.hpp`

**Comment by jyrkive** [comment]
Those names don't sound very good to me. I suggest the following ones instead:

* picture.hpp
* game_version.hpp
* map_settings.hpp
* game_config_manager_tests.hpp (this one is fine)

**Comment by hrubymar10** [comment]
Ok, I'll rename them before commiting it

**Comment by hrubymar10** [comment]
Kill me... 😢

Yes, wesnoth is buildable using Xcode 10 now but doesn't work... Whole wesnoth window is blurry and when you switch to fullscreen you got this:
![image](https://user-images.githubusercontent.com/2218817/43991598-c4af03f4-9d70-11e8-8046-0ee4ac5caed8.png)

(On screenshot you can see my whole display, it is not cropped)

**Comment by Vultraz** [comment]
Looks like some bad DPI-related settings.

**Comment by hrubymar10** [comment]
Yes, see my comments in #3451 PR

**Comment by CelticMinstrel** [comment]
I don't like the name "picture.hpp"... can't think of a better one though...

**Comment by hrubymar10** [comment]
Building with Xcode 10 was fixed in PR #3451 and will be hopefully merged soon. Closing this now
